src/core/cgroup-bpf-ctx.c

```diff
@@ -0,0 +1,58 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+#include "cgroup-bpf-ctx.h"
+#include "cgroup.h"
+#include "memory-util.h"
+#include "strv.h"
+#include "string-util.h"
+
+int cgroup_add_pinned_bpf_program(CGroupContext *c, enum bpf_attach_type attach_type, const char *bpffs_path) {
+        _cleanup_free_ CGroupPinnedBPFProgram *p = NULL;
+        _cleanup_free_ char *dup = NULL;
+
+        assert(c);
+        assert(bpffs_path);
+
+        if (cgroup_contains_pinned_bpf_program(c, attach_type, bpffs_path))
+                return log_warning_errno(SYNTHETIC_ERRNO(EEXIST),
+                                "BPF program %s is already specified in cgroup context", bpffs_path);
+
+        dup = strdup(bpffs_path);
+        if (!dup)
+                return log_oom();
+
+        p = new(CGroupPinnedBPFProgram, 1);
+        if (!p)
+                return log_oom();
+
+        *p = (CGroupPinnedBPFProgram) {
+                .attach_type = attach_type,
+                .bpffs_path = TAKE_PTR(dup),
```

Comment:
wouldn't it make more snese to store an fd ref to this inode instead of a path?

Reply:
@poettering 

Coupling this comment with https://github.com/systemd/systemd/pull/13496#discussion_r555370078, are you suggesting to not to store fs paths at all, so `config_parse_bpf_fs_progs` in `load-fragment`  will directly modify unit, bypassing `CGroupContext`?

Also a struct with fd and attach type fields is already [introduced](https://github.com/systemd/systemd/pull/17655/commits/18d08a1346726b409e2b2f18f339b3b03a63d319#diff-a4c4bb375699a1d85ee8c079ec6f06498a134f07b1dd7ee43e02f9111a8beb3bR9-R12) in https://github.com/systemd/systemd/pull/17655, this change will bring these two PRs closer together. I propose `BPFLoadedProgram` as the name and `src/shared/bpf-loaded-program.[ch]` as source paths for both.

Reply:
@poettering 

Storing paths in context is required to for check program uniqueness, see https://github.com/systemd/systemd/pull/13496#discussion_r560495460

Reply:
hmm, you don't need the full path for that, though, just just filename, i.e. the last component of the path?

Reply:
hmm, strike that, people might want to use symlinks from somewhere in /run or so